server/src/main/java/org/elasticsearch/index/mapper/IntervalThrottler.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.mapper;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Throttles tracked operations based on a time interval, restricting them to 1 per N seconds.
+ */
+enum IntervalThrottler {
+    DOCUMENT_PARSING_FAILURE(60);
+
+    static final int MILLISECONDS_IN_SECOND = 1000;
+
+    private final Acceptor acceptor;
+
+    IntervalThrottler(long intervalSeconds) {
+        acceptor = new Acceptor(intervalSeconds * MILLISECONDS_IN_SECOND);
+    }
+
+    /**
+     * @return true if the operation gets accepted, false if throttled.
+     */
+    boolean accept() {
+        return acceptor.accept();
+    }
+
+    // Defined separately for testing.
+    static class Acceptor {
+        private final long intervalMillis;
+        private final AtomicBoolean lastAcceptedGuard = new AtomicBoolean(false);
+        private volatile long lastAcceptedTimeMillis = 0;
+
+        Acceptor(long intervalMillis) {
+            this.intervalMillis = intervalMillis;
+        }
+
+        boolean accept() {
+            final long now = System.currentTimeMillis();
+            // Check without guarding first, to reduce contention.
+            if (now - lastAcceptedTimeMillis > intervalMillis) {
```

Comment:
maybe do a volatile read once? 

```java
final long lastAcceptedTimeMillis = this. lastAcceptedTimeMillis;
```

and reuse that in the rest of the method body?

Reply:
I think we need to check again under guarding, otherwise a few concurrent ops may slip in and get accepted. I don't expect the second check to be a perf bottleneck, it's behind the `compareAndSet` and executed in sequence so the value should be in the CPU caches etc.

Reply:
Besides, the value gets  updated right below so it'll have to be fetched into registers etc.